### PR TITLE
Eth2-to-Near-relay: improvements and bugs fixes

### PR DESCRIPTION
**Pull request by olga24912**
* handle submission of zero headers
* timeout on submission headers to NEAR
* move timeout for fetching beacon states/blocks into configs
* check that Beacon/ETH1/NEAR networks are sync
* check config input format
* submit only finalized blocks mode
* sleep between loops iterations
* use correct sync committee between periods for signature verification
* don't subtract u64 
* find signature slot with enough sync committee bits sum
* check attestations.len > 0
* change beacon API for new one

### CODE REVIEW

File: eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs [hunk 1]
```diff
@@ -95,14 +107,24 @@ impl Eth2NearRelay {
         while !self.terminate {
             iter_id += 1;
             self.set_terminate(iter_id, max_iterations);
+            skip_fail!(self.wait_for_synchronization(), "Fail to get sync status");
 
             info!(target: "relay", "== New relay loop ==");
+            sleep(Duration::from_secs(12));
```
Comment (karim-en) [inline eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs:113]: We should avoid sleeping in two cases:
1. on the first run
2. when there is big gap, so we want to sync it as fast as possible

File: eth2near/eth2near-block-relay-rs/src/execution_block_proof.rs [hunk 2]
```diff
@@ -129,6 +129,8 @@ mod tests {
 
     const TEST_BEACON_BLOCK_ID: u32 = 741888;
     const BEACON_ENDPOINT: &str = "https://lodestar-kiln.chainsafe.io";
+    const TIMEOUT: u64 = 30;
```
Comment (karim-en) [inline eth2near/eth2near-block-relay-rs/src/execution_block_proof.rs:132]: The same here `TIMEOUT_SECONDS`

File: eth2near/eth2near-block-relay-rs/src/execution_block_proof.rs [hunk 3]
```diff
@@ -129,6 +129,8 @@ mod tests {
...
+    const TIMEOUT_STATE: u64 = 1000;
```
Comment (karim-en) [inline eth2near/eth2near-block-relay-rs/src/execution_block_proof.rs:133]: And here `TIMEOUT_STATE_SECONDS`

File: eth2near/eth2near-block-relay-rs/src/last_slot_searcher.rs [hunk 4]
```diff
@@ -409,7 +409,8 @@ mod tests {
     const FIRST_SLOT: u64 = 1099360;
     const SLOT_WITHOUT_BLOCK: u64 = 1099364;
     const RIGHT_BOUND_IN_SLOT_SEARCH: u64 = 1099500;
-
+    const TIMEOUT: u64 = 30;
```
Comment (karim-en) [inline eth2near/eth2near-block-relay-rs/src/last_slot_searcher.rs:412]: Also please add a suffix here

File: eth2near/eth2near-block-relay-rs/src/last_slot_searcher.rs [hunk 5]
```diff
@@ -409,7 +409,8 @@ mod tests {
...
+    const TIMEOUT_STATE: u64 = 1000;
```
Comment (karim-en) [inline eth2near/eth2near-block-relay-rs/src/last_slot_searcher.rs:413]: And here

File: eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs [hunk 6]
```diff
@@ -458,6 +476,8 @@ mod tests {
 
     const TEST_BEACON_BLOCK_ID: u32 = 741888;
     const BEACON_ENDPOINT: &str = "https://lodestar-kiln.chainsafe.io";
+    const TIMEOUT: u64 = 30;
```
[the remark this answers is not in the document]
Comment (olga24912) [reply eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs]: fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/800/commits/4aefac2a955839abc6233e6c63efe671be60feef

File: eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs [hunk 7]
```diff
@@ -458,6 +476,8 @@ mod tests {
...
+    const TIMEOUT_STATE: u64 = 1000;
```
[the remark this answers is not in the document]
Comment (olga24912) [reply eth2near/eth2near-block-relay-rs/src/beacon_rpc_client.rs]: fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/800/commits/4aefac2a955839abc6233e6c63efe671be60feef

File: eth2near/eth2near-block-relay-rs/src/execution_block_proof.rs (hunk 2 above)
Comment (olga24912) [reply eth2near/eth2near-block-relay-rs/src/execution_block_proof.rs]: fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/800/commits/4aefac2a955839abc6233e6c63efe671be60feef

File: eth2near/eth2near-block-relay-rs/src/execution_block_proof.rs (hunk 3 above)
Comment (olga24912) [reply eth2near/eth2near-block-relay-rs/src/execution_block_proof.rs]: fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/800/commits/4aefac2a955839abc6233e6c63efe671be60feef

File: eth2near/eth2near-block-relay-rs/src/last_slot_searcher.rs (hunk 4 above)
Comment (olga24912) [reply eth2near/eth2near-block-relay-rs/src/last_slot_searcher.rs]: fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/800/commits/4aefac2a955839abc6233e6c63efe671be60feef

File: eth2near/eth2near-block-relay-rs/src/last_slot_searcher.rs (hunk 5 above)
Comment (olga24912) [reply eth2near/eth2near-block-relay-rs/src/last_slot_searcher.rs]: fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/800/commits/4aefac2a955839abc6233e6c63efe671be60feef

File: eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs (hunk 1 above)
Comment (olga24912) [reply eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs]: fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/800/commits/3b32c83caba171589438821e002e00ef14c9d08e

File: eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs [hunk 8]
```diff
@@ -125,11 +148,25 @@ impl Eth2NearRelay {
                     "Network problems during fetching execution blocks"
                 );
                 self.submit_execution_blocks(headers, current_slot, &mut last_eth2_slot_on_near);
+                sleep(Duration::from_secs(5));
                 self.send_light_client_updates(last_eth2_slot_on_near);
```
[the remark this answers is not in the document]
Comment (olga24912) [reply eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs:152]: I don't get it. Last finalized block on near is known at that point. Do you mean we need to find the "last slot on near"?

File: eth2near/eth2near-block-relay-rs/src/config.rs [hunk 9]
```diff
@@ -53,13 +55,84 @@ pub struct Config {
...
+
+        //check contract_account_id
+        let _contract_account_id: near_sdk::AccountId = self.contract_account_id.parse().unwrap();
+        if near_rpc_client.check_account_exists(&self.contract_account_id).unwrap() == false {
+            panic!("Contract account id doesn't exist on NEAR network");
+        }
+
+        //check dao_contract_account_id
+        if let Some(dao_contract_account_id) = self.dao_contract_account_id.clone() {
+            let _dao_contract_account_id: near_sdk::AccountId = dao_contract_account_id.parse().unwrap();
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/config.rs:108]:
```suggestion
        // check `signer_account_id`
        let _signer_account_id: near_sdk::AccountId = self.signer_account_id.parse().unwrap();
        if near_rpc_client.check_account_exists(&self.signer_account_id).unwrap() == false {
            panic!("Signer account id doesn't exist on NEAR network");
        }

        // check `contract_account_id`
        let _contract_account_id: near_sdk::AccountId = self.contract_account_id.parse().unwrap();
        if near_rpc_client.check_account_exists(&self.contract_account_id).unwrap() == false {
            panic!("Contract account id doesn't exist on NEAR network");
        }

        // check `dao_contract_account_id`
        if let Some(dao_contract_account_id) = self.dao_contract_account_id.clone() {
            let _dao_contract_account_id: near_sdk::AccountId = dao_contract_account_id.parse().unwrap();
```

File: eth2near/eth2near-block-relay-rs/src/config.rs [hunk 10]
```diff
@@ -53,13 +55,84 @@ pub struct Config {
...
+            self.contract_type == "dao" ||
+            self.contract_type == "file") {
+            panic!("Unknown contract type {}", self.contract_type);
+        }
+
+        //check near_network_id
+        if !(self.near_network_id == "mainnet" ||
+             self.near_network_id == "testnet") {
+            panic!("Unknown near network id {}", self.near_network_id);
+        }
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/config.rs:135]:
```suggestion
        // check `network`
        if !(self.network == "main" ||
            self.network == "kiln" ||
            self.network == "ropsten" ||
            self.network == "goerli") {
            panic!("Unknown network {}", self.network);
        }

        // check `contract_type`
        if !(self.contract_type == "near" ||
            self.contract_type == "dao" ||
            self.contract_type == "file") {
            panic!("Unknown contract type {}", self.contract_type);
        }

        // check `near_network_id`
        if !(self.near_network_id == "mainnet" ||
             self.near_network_id == "testnet") {
            panic!("Unknown near network id {}", self.near_network_id);
        }
```

**Review by sept-en — CHANGES_REQUESTED** [review]

**Comment by sept-en** [comment]
@olga24912 please merge with the latest `master`

File: eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs [hunk 11]
```diff
@@ -19,6 +22,8 @@ macro_rules! skip_fail {
             Ok(val) => val,
             Err(e) => {
                 warn!(target: "relay", "{}. Error: {}", $msg, e);
+                trace!(target: "relay", "Sleep 12 secs before next loop");
+                sleep(Duration::from_secs(12));
```
[the remark this answers is not in the document]
Comment (olga24912) [reply eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs]: fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/800/commits/3708dbec3a19fc948461d191d280025ef10fe0f2

File: eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs [hunk 12]
```diff
@@ -125,11 +148,25 @@ impl Eth2NearRelay {
                     "Network problems during fetching execution blocks"
                 );
                 self.submit_execution_blocks(headers, current_slot, &mut last_eth2_slot_on_near);
+                sleep(Duration::from_secs(5));
```
[the remark this answers is not in the document]
Comment (olga24912) [reply eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs]: fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/800/commits/3708dbec3a19fc948461d191d280025ef10fe0f2

File: eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (olga24912) [reply eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs]: fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/800/commits/5ada70e4823482804bf33e135dcce26da3cc7851

**Review by karim-en — APPROVED** [review]

File: eth2near/eth2near-block-relay-rs/src/eth1_rpc_client.rs [hunk 13]
```diff
@@ -53,6 +53,29 @@ impl Eth1RPCClient {
...
+            .text()?;
+
+        let val: Value = serde_json::from_str(&res)?;
+        let is_sync = val["result"].as_bool();
+        if let Some(is_sync_val) = is_sync {
+            return Ok(is_sync_val);
+        } else {
+            return Ok(true);
+        }
+    }
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/eth1_rpc_client.rs:78]:
```suggestion
        if let Some(is_sync_val) = is_sync {
            return Ok(is_sync_val);
        }
        
        Ok(true)
    }
```

File: eth2near/eth2near-block-relay-rs/src/hand_made_finality_light_client_update.rs [hunk 14]
```diff
@@ -22,6 +23,12 @@ impl HandMadeFinalityLightClientUpdate {
         attested_slot: u64,
         include_next_sync_committee: bool,
     ) -> Result<LightClientUpdate, Box<dyn Error>> {
+        let (attested_slot, signature_slot) =
+            Self::get_attested_slot_with_enough_sync_committee_bits_sum(
+                &beacon_rpc_client,
+                attested_slot,
+            )?;
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/hand_made_finality_light_client_update.rs:30]:
```suggestion
        let (attested_slot, signature_slot) =
            Self::get_attested_slot_with_enough_sync_committee_bits_sum(
                beacon_rpc_client,
                attested_slot,
            )?;
```

File: eth2near/eth2near-block-relay-rs/src/hand_made_finality_light_client_update.rs [hunk 15]
```diff
@@ -65,17 +79,73 @@ impl HandMadeFinalityLightClientUpdate {
...
+                .sum();
+            if sync_committee_bits_sum * 3 < (64 * 8 * 2) {
+                current_attested_slot = signature_slot;
+                continue;
+            }
+
+            if signature_beacon_body.attestations().len() == 0 {
+                current_attested_slot = signature_slot;
+                continue;
+            }
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/hand_made_finality_light_client_update.rs:127]:
```suggestion
            if signature_beacon_body.attestations().is_empty() {
                current_attested_slot = signature_slot;
                continue;
            }
```

File: eth2near/eth2near-block-relay-rs/src/hand_made_finality_light_client_update.rs [hunk 16]
```diff
@@ -65,17 +79,73 @@ impl HandMadeFinalityLightClientUpdate {
...
+
+            if let Err(err) = beacon_rpc_client
+                .get_beacon_block_header_for_block_id(&format!("{}", current_attested_slot))
+            {
+                if let Some(_) = err.downcast_ref::<NoBlockForSlotError>() {
+                    current_attested_slot = signature_slot;
+                    continue;
+                }
+                return Err(err);
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/hand_made_finality_light_client_update.rs:142]:
```suggestion
                if err.downcast_ref::<NoBlockForSlotError>().is_some() {
                    current_attested_slot = signature_slot;
                    continue;
                }
                return Err(err);
```

File: eth2near/eth2near-block-relay-rs/src/near_rpc_client.rs [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+        let res = self
+            .client
+            .post(&self.endpoint_url)
+            .json(&json_value)
+            .send()?
+            .text()?;
+
+        let val: Value = serde_json::from_str(&res)?;
+
+        return Ok(val["result"].is_object());
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/near_rpc_client.rs:39]:
```suggestion
        Ok(val["result"].is_object())
```

File: eth2near/eth2near-block-relay-rs/src/near_rpc_client.rs [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+            .send()?
+            .text()?;
+
+        let val: Value = serde_json::from_str(&res)?;
+
+        if let Some(is_sync) = val["result"]["sync_info"]["syncing"].as_bool() {
+            return Ok(is_sync);
+        }
+
+        return Ok(true);
```
Comment (sept-en) [inline eth2near/eth2near-block-relay-rs/src/near_rpc_client.rs:63]:
```suggestion
        Ok(true)
```

**Review by sept-en — CHANGES_REQUESTED** [review]

**Review by sept-en — APPROVED** [review]

**Pull request merged**